hack/README.md

```diff
@@ -0,0 +1,215 @@
+
+Example:
+```sh
+#!/bin/bash
+
+# This file contains helpful aliases for manipulating the output text to the terminal as
+# well as functions for one-command augmented printing.
+```
+
+### Function Comments
+Any function that is either non-obvious or long should be documented. Any function in a library should be documented.
+
+All function comments must contain:
+ * the full name of the function (including namespaces) as the first entry on the first line of the function
+ * a description of the function
+ * a list of global variables used in the function
+ * a list of arguments to the function
+ * what the function returns, if it returns something other than the exit status of the last command run
+
+Example:
+```sh
+# os::cmd::internal::expect_exit_code_run_grep runs a command and makes an assertion about its exit code
+# and the contents of the command's output to 'stdout' and/or 'stderr' Output from the command is suppressed
+# unless either `VERBOSE=1` or the test fails. 
+# Globals:
+#  - VERBOSE
+# Arguments:
+#  - 1: the command to run
+#  - 2: the command result code evaluation function
+#  - 3: the arguments to pass to 'grep'
+#  - 4: the command output evaluation function
+# Return:
+#  0 if all assertions are met for the command
+#  1 if any assertion is not met for he command
+```
+
+### Implementation Comments
+Comments should appear for any code that is difficult to understand, techincal, non-obvious or otherwise interesting.
+
+## Formatting
+
+### Indentation
+All files will be indented four spaces, no tabs. Alignment is used often for multi-line command and will not be consistent across developer workstations if tabs are used.
+
+### Line Length
+Lines should not be longer than 120 characters. If lines are longer than this, determine if it is the result of long pipelines and re-format as specified [below](#pipelines).
+
+### Pipelines
+Pipelines should be formatted one per line if they don't all fit into one line. If the pipeline fits on one line, leave it as one line. These rules apply to chains of commands linked with the pipe `|` or logical statements `&&` or `||`.
+
+Example:
+```sh
+# Pipeline fits on one line
+command1 | command2 | command3
+
+# Pipeline doesn't fit on one line
+command1 | \
+    | command2 \
+    | command3 \
+    | command4
+```
+
+### Loops
+Put `; then` or `; do` on the same line as the preceding `for`, `while`, or `if`. `else` should be on a separate line. Concluding statements `done` and `fi` should be on their own lines and aligned with the opening statements.
+
+Example:
+```sh
+for item in "${items}"; do
+    if [[ -d "${item}" ]]; then
+        some_action "${item}"
+    else
+        some_other_action "${item}"
+    fi
+done
+```
+
+### Variable Expansion
+Quote variables, prefer `"${var}"` when possible
+
+Don't brace-quote single-character shell special characters or positional parameters unless necessary or if the brace-quotes aids in readability.
```

Comment:
I believe the grouping is ((single-character (((shell special) characters) or positional)) parameters).

The phrase "single-character positional or special parameters" would more closely match the nomenclature in the manpage, and maybe it would be a little clearer.